crates/bevy_app/src/plugin.rs

```diff
@@ -56,19 +89,46 @@ use std::any::Any;
 /// # fn damp_flickering() {}
 /// ````
 pub trait Plugin: Downcast + Any + Send + Sync {
-    /// Configures the [`App`] to which this plugin is added.
-    fn build(&self, app: &mut App);
+    /// Returns required sub apps before finalizing this plugin.
+    fn require_sub_apps(&self) -> Vec<InternedAppLabel> {
+        Vec::new()
+    }
 
-    /// Has the plugin finished its setup? This can be useful for plugins that need something
-    /// asynchronous to happen before they can finish their setup, like the initialization of a renderer.
-    /// Once the plugin is ready, [`finish`](Plugin::finish) should be called.
-    fn ready(&self, _app: &App) -> bool {
+    /// Pre-configures the [`App`] to which this plugin is added. This is usually executed before
+    /// the event loop is started.
+    fn init(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Is the plugin ready to be built?
+    fn ready_to_build(&self, _app: &mut App) -> bool {
```

Comment:
Shouldn't this take a `&App` rather than a `&mut App`? This should only be reading the world / app state.

Reply:
Ugh, right. Okay. This can stay for now since https://github.com/bevyengine/bevy/issues/3774 is not fixed.